### PR TITLE
E2e protocol ricardo

### PR DESCRIPTION
**Pull request by corrideat**

### CODE REVIEW

File: frontend/controller/actions/identity.js [hunk 1]
```diff
@@ -369,24 +360,128 @@ export default (sbp('sbp/selectors/register', {
...
 
+    // Used to store stateEncryptionKeyId, salt, encryptedStateEncryptionKey
+    const extraLoginAttributes = {}
+
     try {
-      sbp('appLogs/startCapture', username)
-      const state = await sbp('gi.db/settings/load', username)
-      let contractIDs = []
+      // Utility class to represent an empty state during error handling
+      class EmptyState extends Error {}
```
Comment (taoeffect) [inline frontend/controller/actions/identity.js]: Creating classes in the middle of a function is not clean code.

If you must create a new class for the error, please put it in the appropriate place along with the other `GIError*` classes

File: frontend/controller/actions/identity.js [hunk 2]
```diff
@@ -369,24 +360,128 @@ export default (sbp('sbp/selectors/register', {
...
+        const stateKeyEncryptionKey = await deriveKeyFromPassword(CURVE25519XSALSA20POLY1305, password, salt + stateEncryptionKeyId)
+
+        // Once everything is place, set the attributes for extraLoginAttributes
+        extraLoginAttributes.stateEncryptionKeyId = stateEncryptionKeyId
+        extraLoginAttributes.salt = salt
+        extraLoginAttributes.encryptedStateEncryptionKey = encrypt(stateKeyEncryptionKey, stateEncryptionKeyS, stateEncryptionKeyId)
+
+        // Save the state encryption key to local storage
+        await sbp('gi.db/settings/save', stateEncryptionKeyId, stateEncryptionKeyS)
+      })
```
Comment (taoeffect) [inline frontend/controller/actions/identity.js]: Please keep the login function as simple as it was before by moving all of this code elsewhere.

File: frontend/model/state.js [hunk 3]
```diff
@@ -99,9 +100,22 @@ sbp('sbp/selectors/register', {
...
+      const stateEncryptionKeyS = await sbp('gi.db/settings/load', stateEncryptionKeyId)
+      if (!stateEncryptionKeyS) throw new Error(`Unable to retrieve the key corresponding to key ID ${stateEncryptionKeyId}`)
+      // Encrypt the current state
+      const encryptedState = encrypt(stateEncryptionKeyS, JSON.stringify(state), identityContractID)
+      // Save the three fields of the encrypted state:
+      //   (1) stateEncryptionKeyId
+      //   (2) salt
+      //   (3) encryptedStateEncryptionKey (used for recovery when re-logging in)
+      //   (4) encryptedState
+      await sbp('gi.db/settings/save', identityContractID, `${stateEncryptionKeyId}.${salt}.${encryptedStateEncryptionKey}.${encryptedState}}`)
```
Comment (taoeffect) [inline frontend/model/state.js]: Please move all of this into `gi.db/settings/save`

File: frontend/model/state.js [hunk 4]
```diff
@@ -99,9 +100,22 @@ sbp('sbp/selectors/register', {
     // IMPORTANT! DO NOT CALL VUEX commit() in here in any way shape or form!
     //            Doing so will cause an infinite loop because of store.subscribe below!
     if (state.loggedIn) {
+      const { identityContractID, stateEncryptionKeyId, salt, encryptedStateEncryptionKey } = state.loggedIn
       state.notifications = applyStorageRules(state.notifications || [])
-      // TODO: encrypt this
-      await sbp('gi.db/settings/save', state.loggedIn.username, state)
```
Comment (taoeffect) [inline frontend/model/state.js:104]: Please preserve this

**Review by taoeffect — CHANGES_REQUESTED** [review]
Please preserve the original code in `state.js` and `actions/identity.js` as much as possible by moving all of the encryption code into `gi.db/settings` related selectors

File: frontend/controller/actions/identity.js (hunk 2 above)
Comment (taoeffect) [reply frontend/controller/actions/identity.js]: Can be put in a new selector called `'gi.db/settings/loadEncrypted'` (and create a `'gi.db/settings/saveEncrypted'` below)

File: frontend/controller/actions/identity.js [hunk 5]
```diff
@@ -62,16 +62,6 @@ export default (sbp('sbp/selectors/register', {
...
-    // TODO: make sure we namespace these names:
-    //       https://github.com/okTurtles/group-income/issues/598
-    const oldSettings = await sbp('gi.db/settings/load', username)
-    if (oldSettings) {
-      // TODO: prompt to ask user before deleting and overwriting an existing user
-      //       https://github.com/okTurtles/group-income/issues/599
-      console.warn(`deleting settings for pre-existing identity ${username}!`, oldSettings)
-      await sbp('gi.db/settings/delete', username)
-    }
-
```
Comment (taoeffect) [inline frontend/controller/actions/identity.js:74]: Why was this removed?

File: frontend/controller/actions/identity.js [hunk 6]
```diff
@@ -350,11 +340,12 @@ export default (sbp('sbp/selectors/register', {
...
     }
   },
-  'gi.actions/identity/login': async function ({ username, password }: {
-    username: string, password: ?string
+  'gi.actions/identity/login': async function ({ username, password, identityContractID }: {
+    username: ?string, password: ?string, identityContractID: ?string
   }) {
-    // TODO: Insert cryptography here
-    const identityContractID = await sbp('namespace/lookup', username)
+    if (username) {
```
Comment (taoeffect) [inline frontend/controller/actions/identity.js:346]: Under what circumstances would `username` not be set and why?

Please add a comment explaining

File: frontend/controller/actions/identity.js (hunk 5 above)
Comment (corrideat) [reply frontend/controller/actions/identity.js:74]: No longer needed as user settings are now indexed by contract ID, so there are no possible conflicts when creating a new account.

File: frontend/controller/actions/identity.js (hunk 6 above)
Comment (corrideat) [reply frontend/controller/actions/identity.js:346]: When the page is refreshed (or otherwise reopened on an active session), and because now the values are indexed by username.

File: frontend/controller/actions/identity.js [hunk 7]
```diff
@@ -370,23 +361,53 @@ export default (sbp('sbp/selectors/register', {
...
+      const loginAttributes: Object = { identityContractID, encryptionParams }
+
+      // If username was not provided, retrieve it from the state
+      // TODO: This is temporary until username is no longer needed internally
+      // in contracts
+      if (username) {
+        loginAttributes.username = username
+      } else {
+        loginAttributes.username = Object.entries(state.namespaceLookups).find(([k, v]) => v === identityContractID)?.[0]
+      }
```
Comment (taoeffect) [inline frontend/controller/actions/identity.js]: This can be simplified:

```js
const loginAttributes = { identityContractID, encryptionParams, username }
if (!loginAttributes.username) {
  loginAttributes.username = Object.entries(state.namespaceLookups).find(([k, v]) => v === identityContractID)?.[0]
}
```

File: frontend/model/state.js (hunk 4 above)
Comment (corrideat) [reply frontend/model/state.js:104]: ?
